pageserver/src/tenant/block_io.rs

```diff
@@ -76,6 +68,30 @@ impl<'a> Deref for BlockLease<'a> {
     }
 }
 
+pub(crate) enum BlockReaderRef<'a> {
+    FileBlockReaderVirtual(&'a FileBlockReader<VirtualFile>),
+    FileBlockReaderFile(&'a FileBlockReader<std::fs::File>),
+    EphemeralFile(&'a EphemeralFile),
+    Adapter(Adapter<&'a DeltaLayerInner>),
+    #[cfg(test)]
+    TestDisk(&'a super::disk_btree::tests::TestDisk),
+}
+
+impl<'a> BlockReaderRef<'a> {
+    #[inline(always)]
```

Comment:
Do you know if this makes any difference? I could imagine it is helpful with these kinds of structures where there's a lot of fan out but I'd imagine llvm seeing through to the only choice, but is it unable to do that without this hint?

Reply:
Yeah as we don't have any `dyn BlockReader` cases on master, all of the cases where `BlockReaderRef` is used is in cases where we know the concrete type that implements `BlockReader`, and thus we also know the implementation of the `block_cursor` function, which always creates a specific `BlockReaderRef`. In other words, `BlockReaderRef::read_blk` will always only resolve to one specific enum variant at each call site.

Ideally LLVM realizes this as well and spares the dispatch overhead in `read_blk` (which is non-trivial due to the (upcoming) async state machines). For this to happen, LLVM needs insight into the functions involved. Many LLVM optimization passes are implemented only to apply to the same function, and if LLVM doesn't inline something, that pass won't recognize the case, so that's why I put the inlining hint here.